rpc/filters.go

```diff
@@ -14,30 +15,67 @@ import (
 
 // Filter can be used to retrieve and filter logs.
 type Filter struct {
-	addresses []common.Address
-	topics    [][]common.Hash
-
-	block common.Hash // Block hash if filtering a single block
+	backend            Backend
+	fromBlock, toBlock *big.Int         // start and end block numbers
+	addresses          []common.Address // contract addresses to watch
+	topics             [][]common.Hash  // log topics to watch for
+	blockHash          *common.Hash     // Block hash if filtering a single block
 }
 
-// NewBlockFilter creates a new filter which directly inspects the contents of
-// a block to figure out whether it is interesting or not.
-func NewBlockFilter(block common.Hash, addresses []common.Address, topics [][]common.Hash) *Filter {
-	// Create a generic filter and convert it into a block filter
-	filter := newFilter(addresses, topics)
-	filter.block = block
-	return filter
+// NewFilter returns a new Filter
+func NewFilter(backend Backend, criteria *filters.FilterCriteria) *Filter {
+	return &Filter{
+		backend:   backend,
+		fromBlock: criteria.FromBlock,
+		toBlock:   criteria.ToBlock,
+		addresses: criteria.Addresses,
+		topics:    criteria.Topics,
+	}
 }
 
-// newFilter creates a generic filter that can either filter based on a block hash,
-// or based on range queries. The search criteria needs to be explicitly set.
-func newFilter(addresses []common.Address, topics [][]common.Hash) *Filter {
+// NewFilterWithBlockHash returns a new Filter with a blockHash.
+func NewFilterWithBlockHash(backend Backend, criteria *filters.FilterCriteria) *Filter {
 	return &Filter{
-		addresses: addresses,
-		topics:    topics,
+		backend:   backend,
+		fromBlock: criteria.FromBlock,
+		toBlock:   criteria.ToBlock,
+		addresses: criteria.Addresses,
+		topics:    criteria.Topics,
+		blockHash: criteria.BlockHash,
 	}
 }
 
+// NewBlockFilter creates a new filter that notifies when a block arrives.
+func NewBlockFilter(backend Backend) *Filter {
+	// TODO: finish
```

Comment:
Is this going to be addressed in a later PR?

Reply:
yes, I didn't want to implement too much in this PR, but just outline what the structure would look like. there is an issue open for this